src/webots/nodes/WbEmitter.cpp

```diff
@@ -105,7 +109,35 @@ void WbEmitter::updateRange() {
   mNeedToSetRange = true;
 }
 
+bool WbEmitter::isChannelAllowed() {
+  const int allowedChannelsSize = mAllowedChannels->size();
+  if (allowedChannelsSize > 0) {
+    const int currentChannel = (int)mChannel->value();
+    for (int i = 0; i < allowedChannelsSize; i++) {
+      if (currentChannel == mAllowedChannels->item(i))
+        return true;
+    }
+    return false;
+  }
+  return true;
+}
+
+void WbEmitter::updateAllowedChannels() {
+  if (!isChannelAllowed()) {
+    parsingWarn(
+      tr("'allowedChannels' does not contain current 'channel'. Setting 'channel' to %1.").arg(mAllowedChannels->item(0)));
+    mChannel->setValue(mAllowedChannels->item(0));
+    mNeedToSetChannel = true;
+  }
+
+  mNeedToSetAllowedChannels = true;
+}
+
 void WbEmitter::updateChannel() {
+  if (!isChannelAllowed()) {
+    parsingWarn(tr("'channel' is not included in 'allowedChannels'"));
+    return;
```

Comment:
Here it is still possible to set the channel value to an invalid value because this function simply prints a warning  if the new value is invalid but doesn't reset it.